### PR TITLE
test: jax network components

### LINKED ISSUE #526: [TEST] Jax networks
### What do you want to test?
Jax networks components

### Outline of test structure
* Unit tests
* Test components and hooks

### Definition of done
Passing checks, cover all hooks, edge cases considered

### Mandatory checklist before making a PR
* [ ] The success criteria laid down in “Definition of done” are met.
* [ ] Test code is documented - docstrings for methods and classes, static types for arguments.
* [ ] Documentation is updated - README, CONTRIBUTING, or other documentation.

### PR DESCRIPTION
**Pull request by mmorris44**
## What?
Unit tests for jax networks components
## Why?
Components currently not tested. Needs to be robust for jax Mava release
## How?
- Test `DefaultNetworks` component
## Extras
Closes #526

### CODE REVIEW

File: tests/jax/components/building/networks_test.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        test_builder: Pytest fixture for test system builder
+
+    Returns:
+        None.
+    """
+    assert not hasattr(test_builder.store, "key")
+    test_default_networks.on_building_init_start(test_builder)
+    assert hasattr(test_builder.store, "key")
+    assert isinstance(test_builder.store.key, jax.random.PRNGKeyArray) or isinstance(
+        test_builder.store.key, jax.numpy.DeviceArray
```
Comment (RuanJohn) [inline tests/jax/components/building/networks_test.py:121]: Just checking about the or here. Does the `PRNGKeyArray` change to a `DeviceArray` later on?

Comment (mmorris44) [reply tests/jax/components/building/networks_test.py:121]: There is some weird `jax` behaviour here. The function `jax.random.PRNGKey(self.config.seed)` is supposed to produce a `PRNGKeyArray` according to its type signature, but it doesn't. In practice, it returns a `DeviceArray`.

To make it worse, elsewhere it says it returns a `KeyArray` which is just `KeyArray = Union[Array, prng.PRNGKeyArray]` which is just `Array = Any`. So it's just `Any`, which is useless.

My solution felt the most robust to me. I could just change them to `jax.numpy.DeviceArray` instead?

Comment (RuanJohn) [reply tests/jax/components/building/networks_test.py:121]: Ah, thank you for clarifying. I think your solution is the best in that case.

File: tests/jax/components/building/networks_test.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    """Pytest fixture for system builder.
+
+    Adds mock env specs and agent net keys to store.
+
+    Returns:
+        System builder with no components.
+    """
+    test_builder = Builder(components=[])
+    test_builder.store.environment_spec = make_fake_env_specs()
+    test_builder.store.agent_net_keys = {"key1": "value1", "key2": "value2"}
```
Comment (RuanJohn) [inline tests/jax/components/building/networks_test.py:88]: Is it possible to give these slightly more descriptive names?

File: tests/jax/components/building/networks_test.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    test_default_networks.on_building_init_start(test_builder)
+    networks = test_builder.store.network_factory()
+
+    keys = []
+    for network in networks.values():
+        assert isinstance(network["rng_key"], jax.random.PRNGKeyArray) or isinstance(
+            network["rng_key"], jax.numpy.DeviceArray
+        )
+        keys.append(tuple(network["rng_key"].tolist()))
+    assert len(set(keys)) == 1
```
Comment (RuanJohn) [inline tests/jax/components/building/networks_test.py:188]: Not sure about the key handling, but does this mean that all networks should always share the same `rng_key`?

**Comment by RuanJohn** [comment]
Thank you @mmorris44! Looks good, just a few small questions from my side.

File: tests/jax/components/building/networks_test.py (hunk 2 above)
Comment (mmorris44) [reply tests/jax/components/building/networks_test.py:72]: Agreed. Done.

File: tests/jax/components/building/networks_test.py (hunk 3 above)
Comment (mmorris44) [reply tests/jax/components/building/networks_test.py:188]: Yes it does. I have added a comment to make this clearer.

**Review by RuanJohn — APPROVED** [review]
Thank you @mmorris44. Looks good to me :call_me_hand:

File: tests/jax/components/building/networks_test.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+@pytest.fixture
+def test_network_factory() -> Callable:
+    """Pytest fixture for network factory.
+
+    Returns:
+        Network factory using custom make_default_networks.
+    """
+
+    def make_default_networks(
```
Comment (KaleabTessera) [inline tests/jax/components/building/networks_test.py:22]: @mmorris44 Can't we use the `make_default_networks` that exist in ppo already?

Comment (mmorris44) [reply tests/jax/components/building/networks_test.py:22]: Thanks for the review @KaleabTessera. Yes we could, but I don't want the tests to be using something external to the code I'm trying to test here. I.e. if someone changes the `make_default_networks` of PPO, these tests should not break. Thus, I defined my own version here to show how the function might be defined in the wild.

Comment (KaleabTessera) [reply tests/jax/components/building/networks_test.py:22]: Cool. Makes sense.

File: tests/jax/components/building/networks_test.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        test_builder: Pytest fixture for test system builder
+
+    Returns:
+        None.
+    """
+    test_default_networks.on_building_init_start(test_builder)
+    networks = test_builder.store.network_factory()
+
+    for network in networks.values():
+        assert network["environment_spec"]._keys == ["agent_0", "agent_1"]
```
Comment (KaleabTessera) [inline tests/jax/components/building/networks_test.py:144]: Can we please get the `["agent_0", "agent_1"]` values from `make_fake_env_specs()` directly. It is possible someone changes the default values in `make_fake_env_specs()` and then this test fails.

File: tests/jax/components/building/networks_test.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        assert network["environment_spec"]._specs["agent_0"].observations.shape == (
```
Comment (KaleabTessera) [inline tests/jax/components/building/networks_test.py:145]: Same as above.

File: tests/jax/components/building/networks_test.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+    Returns:
+        None.
+    """
+    test_default_networks.on_building_init_start(test_builder)
+    networks = test_builder.store.network_factory()
+
+    for network in networks.values():
+        assert network["agent_net_keys"] == {
+            "net_key1": "network_1",
```
Comment (KaleabTessera) [inline tests/jax/components/building/networks_test.py:168]: Same as above.

Comment (mmorris44) [reply tests/jax/components/building/networks_test.py:168]: I defined these myself in this file, so it does not apply here. See function `test_builder()` on line 73/74.

File: tests/jax/components/building/networks_test.py (hunk 5 above)
Comment (mmorris44) [reply tests/jax/components/building/networks_test.py:144]: Done. Thanks for spotting this.

File: tests/jax/components/building/networks_test.py (hunk 6 above)
Comment (mmorris44) [reply tests/jax/components/building/networks_test.py:145]: Done.

File: tests/jax/components/building/networks_test.py (hunk 3 above)
Comment (mmorris44) [reply tests/jax/components/building/networks_test.py:199]: See #592. Will address in a separate PR

**Review by KaleabTessera — APPROVED** [review]
Thanks @mmorris44 :rocket: !

**Review by RuanJohn — APPROVED** [review]
Thanks @mmorris44 :)
